internal/server/v2/observation/calculation_util.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package observation
+
+import (
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"reflect"
+	"strings"
+
+	pb "github.com/datacommonsorg/mixer/internal/proto"
+)
+
+// The info of a node in the AST tree.
+type ASTNode struct {
+	StatVar string
+	Facet   *pb.Facet
+}
+
+type VariableFormula struct {
+	Expr ast.Expr
+	// Key is encodeForParse(<name of ASTNode>).
+	NodeDataMap map[string]*ASTNode
+	StatVars    []string
+}
+
+// Golang's AST package is used for parsing the formula, so we need to avoid sensitive tokens for
+// AST. For those tokens, we swap them with insensitive tokens before the parsing, then swap them
+// back after the parsing.
+var (
+	encodeForParseTokenMap = map[string]string{
+		"dc/":          "_DC_SLASH_",
+		"dcAggregate/": "_DC_AGGREGATE_SLASH_",
+		"[":            "_LEFT_SQUARE_BRACKET_",
+		"]":            "_RIGHT_SQUARE_BRACKET_",
+		"=":            "_EQUAL_TO_",
+		";":            "_SEMICOLON_",
+	}
+	decodeForParseTokenMap = map[string]string{
+		"_DC_SLASH_":             "dc/",
+		"_DC_AGGREGATE_SLASH_":   "dcAggregate/",
+		"_LEFT_SQUARE_BRACKET_":  "[",
+		"_RIGHT_SQUARE_BRACKET_": "]",
+		"_EQUAL_TO_":             "=",
+		"_SEMICOLON_":            ";",
```

Comment:
instead of write the reverse map manually, how about reverting encodeForParseTokenMap?

Reply:
Good idea, done